### PR TITLE
8336860: x86: Change integer src operand for CMoveL of 0 and 1 to long

### PR DESCRIPTION
**Pull request by jaskarth**
Hi all,
This patch fixes `cmovL_imm_01*` instructions matching against an integer immediate 1 instead of a long immediate 1. I noticed while looking at the backend implementation of CMove that the rules specify `immI_1` instead of `immL1`, which means that the instructions can't be matched and instead falls through to the base case. I added a small benchmark and got these results:
```c
                                        Baseline                     Patch           Improvement
Benchmark                Mode  Cnt    Score   Error  Units      Score   Error  Units
BasicRules.cmovL_imm_01  avgt   15  259.073 ± 5.806  ns/op    231.108 ± 2.730  ns/op  (+ 11.41%)
```
Thoughts and reviews would be appreciated!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336860](https://bugs.openjdk.org/browse/JDK-8336860): x86: Change integer src operand for CMoveL of 0 and 1 to long (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) Review applies to [586d7703](https://git.openjdk.org/jdk/pull/20275/files/586d770367a42a7387a09957ab0cc89b5c9dd5d5)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) Review applies to [02f8d0f7](https://git.openjdk.org/jdk/pull/20275/files/02f8d0f7fb8c51415b5d96f5092b36f1cbac4b49)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**) Review applies to [1b926ecf](https://git.openjdk.org/jdk/pull/20275/files/1b926ecf1dbc3b1bc905a186b0ed3082a5c4211b)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20275/head:pull/20275` \
`$ git checkout pull/20275`

Update a local copy of the PR: \
`$ git checkout pull/20275` \
`$ git pull https://git.openjdk.org/jdk.git pull/20275/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20275`

View PR using the GUI difftool: \
`$ git pr show -t 20275`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20275.diff">https://git.openjdk.org/jdk/pull/20275.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20275#issuecomment-2241979689)

### CODE REVIEW

**Comment by liach** [comment]
I think you should put the benchmark somewhere in `test/micro/org/openjdk/bench/vm/compiler` instead of `java/lang`; that's why this is labeled "core-libs" by the bot.

**Comment by jaskarth** [comment]
Thanks for the heads up! I put them there since the change that originally introduced the code put its benchmarks there, but I think it would make sense to move it to `bench/vm/compiler`.

**Review by shipilev — APPROVED** [review]
Ah, oops. @merykitty might confirm this was not intentional.

**Review by merykitty — APPROVED** [review]
Yes that was my mistake, thanks a lot for noticing and fixing

**Comment by jaskarth** [comment]
Thank you for your reviews, @shipilev and @merykitty!

I've moved the benchmark and updated the performance results accordingly.

**Review by shipilev — APPROVED** [review]
Ah, stuck on needing the re-review? Yes, still good.

**Comment by jaskarth** [comment]
Oh, thank you :)
Should I wait for someone to run testing on the patch?

**Comment by liach** [comment]
@eme64 or @TobiHartmann might take a look too, I guess?

**Comment by eme64** [comment]
@liach @jaskarth I'll run some testing.

Can you point me to the "base case" you mention in your PR description?

**Comment by jaskarth** [comment]
Thanks for running testing @eme64!
The base case I mention is this:
https://github.com/openjdk/jdk/blob/34ee06f51122e8afb875cc3b75f513912272fd9b/src/hotspot/cpu/x86/x86_64.ad#L6405-L6415
Instead of matching the flow-free `SETcc` idiom, we fall back to using the default `cmovq` behavior.

**Review by eme64 — APPROVED** [review]
Changes look good.

Could there be an IR test here? Matching on the opto assembly?

**Comment by jaskarth** [comment]
I think IR test matching the opto assembly output is a good idea! I'll add one in the next commit.

**Comment by jaskarth** [comment]
@eme64 I've added an IR test that matches all of the relevant comparison types.

**Review by jatin-bhateja — APPROVED** [review]
LGTM

File: test/hotspot/jtreg/compiler/c2/irTests/CMoveLConstants.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+package compiler.c2.irTests;
+
+import jdk.test.lib.Asserts;
+import compiler.lib.ir_framework.*;
+
+/*
+ * @test
+ * bug 8336860
+ * @summary Verify codegen for CMoveL with constants 0 and 1
+ * @requires os.simpleArch == "x64"
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/c2/irTests/CMoveLConstants.java:33]: Hmm. How about only requiring `x64` in the IR rules, but running the test on all platforms? That way we can at least check correct results.

**Review by eme64 — APPROVED** [review]
Otherwise, it looks good to me!

**Review by eme64 — APPROVED** [review]
I'm now good with it, but I'm away from my work laptop and cannot run our testing...

@chhagedorn @TobiHartmann Can one of you please run testing for this before it is integrated?

**Review by chhagedorn — APPROVED** [review]
Testing looked good!

**Comment by jaskarth** [comment]
Thank you for the testing @chhagedorn, and thanks everyone for the reviews!

/integrate

**Pull request closed without merging**